tests/cluster/tests/29-slot-migration-response.tcl

```diff
@@ -0,0 +1,66 @@
+# Tests for the response of slot migrations.
+
+source "../tests/includes/init-tests.tcl"
+source "../tests/includes/utils.tcl"
+
+# TODO: This test currently tests the ASK and MOVED replies, need to add a test to check for the UNSTABLE reply
+
+test "Create a 10 nodes cluster" {
+    create_cluster 10 0
+    config_set_all_nodes cluster-allow-replica-migration no
+}
+
+test "Cluster is up" {
+    assert_cluster_state ok
+}
+
+set cluster [redis_cluster 127.0.0.1:[get_instance_attrib redis 0 port]]
+catch {unset nodefrom}
+catch {unset nodeto}
+
+$cluster refresh_nodes_map
+
+test "Set many keys" {
+    for {set i 0} {$i < 40000} {incr i} {
+        $cluster set $i $i
+    }
+}
+
+test "Keys are accessible" {
+    for {set i 0} {$i < 40000} {incr i} {
+        assert { [$cluster get $i] eq $i }
+    }
+}
+
+test "Migration of slot x" {
+
+    set slot 10
+    array set nodefrom [$cluster masternode_for_slot $slot]
+    array set nodeto [$cluster masternode_notfor_slot $slot]
+
+
+    $nodeto(link) cluster setslot $slot importing $nodefrom(id)
+    $nodefrom(link) cluster setslot $slot migrating $nodeto(id)
+
+    # Get a key from that slot
+    set key [$nodefrom(link) cluster GETKEYSINSLOT $slot "1"]
+
+    # MOVED REPLY
+    catch {[$nodeto(link) set $key "newVal"]} f
+    if {![string match "*MOVED*" $f]} {
+        fail "Should contain MOVED when migrating"
+    }
+
+    # ASK REPLY
+    catch {[$nodefrom(link) set "abc{$key}" "newVal"]} g
+    if {![string match "*ASK*" $g]} {
+        fail "Should contain ASK when migrating"
+    }
+
+    # TODO UNSTABLE REPLY
+    # catch {[$nodefrom(link) mset "a{$key}" "newVal" "abc{$key}" "newVal2"]} f
+    # puts $f
+    # if {![string match "*UNSTABLE*" $f]} {
+    #     fail "Should contain ASK when migrating"
+    # }
```

Comment:
The Cluster Spec page says: "Operations on keys that don’t exist or are - during the resharding - split between the source and destination nodes, will generate a `-TRYAGAIN` error."

You probably need one key which doesn't exist (`"abc{$key}"`) and one which does exist (`$key`, not `"a{$key}"`).

Maybe we also want to test `ASKING; MSET $key "newVal" "abc{$key}" "newVal2"` to the importing node too. It should also return a -TRYAGAIN error if one key (or both?) doesn't exist.

Reply:
Hello @zuiderkwast , thanks for the suggestion. In the case you mentioned "one key which doesn't exist ("abc{$key}") and one which does exist ($key, not "a{$key}")." I get the ask reply instead of the -TRYAGAIN error.

Reply:
That's strange. According to "Operations on keys that don’t exist or are - during the resharding - split between the source and destination nodes, will generate a -TRYAGAIN error." it seems that if one key is missing or both keys are missing should both result in -TRYAGAIN.

Reply:
Yes, that is what I thought as well. But in both scenarios, when one key is missing or when both are missing then I am getting the ASK reply.

Reply:
Ahh! :zap: I can see that now in the implementation in cluster.c:

```C
    /* If we don't have all the keys and we are migrating the slot, send
     * an ASK redirection. */
    if (migrating_slot && missing_keys) {
        if (error_code) *error_code = CLUSTER_REDIR_ASK;
        return server.cluster->migrating_slots_to[slot];
    }
```

The above code doesn't care about the `multiple_keys` variable. I think if all keys are missing it makes sense to send ASK, but if at least one key is still in the migrating node, it means that the importing node doesn't have it yet.

The TRYAGAIN error is only sent for the importing node after ASKING:

```C
    /* If we are receiving the slot, and the client correctly flagged the
     * request as "ASKING", we can serve the request. However if the request
     * involves multiple keys and we don't have them all, the only option is
     * to send a TRYAGAIN error. */
    if (importing_slot &&
        (c->flags & CLIENT_ASKING || cmd->flags & CMD_ASKING))
    {
        if (multiple_keys && missing_keys) {
            if (error_code) *error_code = CLUSTER_REDIR_UNSTABLE;
            return NULL;
        } else {
            return myself;
        }
    }
```

So what can we do? Shall we consider it a bug and fix it or shall we change the documentation and tests to follow the implementation?

Reply:
I think both options are valid here. @madolson , do you have any comments about this? Thanks.

Reply:
This is fixed here I believe? https://github.com/redis/redis/pull/9526/files

Reply:
Very nice. With that fix, the commented test should work (if we check for "TRYAGAIN" instead of "UNSTABLE").

I think we should also test:
* that the target can return -TRYAGAIN, if ASKING is sent and some but not all keys exist.
* that the target returns -MOVED if ASKING is not sent.

Reply:
@hwware Ok, the other change has been merged, this test should work correctly now.